File: Day03/d3.hs
import System.IO
import Data.List
import Data.Char

main = do
    input <- lines <$> readFile "input.txt"
    print $  part1 input
    print $  part2 input

part1 input = sum $ score <$> map (\x -> head $ uncurry intersect $ splitAt (length x `div` 2) x) input
part2 input 
    | null input = 0
    | otherwise = score (head $ foldl1 intersect $ take 3 input) + part2 (drop 3 input)

score c 
    | isLower c = ord c - ord 'a' + 1
    | otherwise = ord c - ord 'A' + 27

File: Day03/d3.py
with open("input.txt", "r") as inputFile:
    lines = inputFile.read().splitlines()

    getPriorities = lambda l: sum(ord(c) - (ord('a') - 1 if c.islower() else ord('A') - 27) for c in l)
    intersect = lambda a, b, c = "": set(a).intersection(set(b)).intersection(set(c if c != "" else b)).pop()

    print(getPriorities([intersect(line[:len(line) // 2], line[len(line) // 2:]) for line in lines]))
    print(getPriorities([intersect(*lines[ptr: ptr + 3]) for ptr in range(0, len(lines), 3)]))